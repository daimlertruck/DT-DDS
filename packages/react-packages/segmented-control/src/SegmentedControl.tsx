import { BaseProps } from '@dt-dds/react-core';
import { ComponentPropsWithoutRef } from 'react';

import { SegmentedControlOption } from './components/SegmentedControlOption';
import { SegmentedControlProvider } from './context';
import { SegmentedControlStyled } from './SegmentedControl.styled';

export interface SegmentedControlProps
  extends ComponentPropsWithoutRef<'div'>,
    BaseProps {
  selectedValue?: string | undefined;
  showLabel?: boolean;
  onChangeControl?: (value: string) => void;
}

const SegmentedControl = ({
  dataTestId,
  children,
  showLabel = false,
  selectedValue,
  onChangeControl = () => null,
  ...props
}: SegmentedControlProps) => {
  return (
    <SegmentedControlProvider
      onChange={onChangeControl}
      selectedValue={selectedValue}
      showLabel={showLabel}
    >
      <SegmentedControlStyled data-testid={dataTestId} role='group' {...props}>
        {children}
      </SegmentedControlStyled>
    </SegmentedControlProvider>
  );
};

SegmentedControl.Option = SegmentedControlOption;

export { SegmentedControl };
