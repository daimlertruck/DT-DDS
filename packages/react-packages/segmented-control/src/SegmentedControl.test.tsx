import { withProviders } from '@dt-dds/react-core';
import { render, fireEvent } from '@testing-library/react';

import { SegmentedControl } from './SegmentedControl';

const ProvidedSegmentedControl = withProviders(SegmentedControl);

describe('SegmentedControl', () => {
  it('renders without errors', () => {
    const { getByTestId, getByRole, container } = rendersSegmentedControl();

    const control = getByTestId('segmented-control');

    fireEvent.click(getByRole('button', { name: 'Gallery' }));
    expect(control).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('calls handleChange handler when an option is clicked', () => {
    const handleChange = jest.fn();
    const { getByRole } = rendersSegmentedControl(handleChange);

    fireEvent.click(getByRole('button', { name: 'Gallery' }));
    expect(handleChange).toHaveBeenCalledWith('gallery');
  });

  it('renders options with tooltips when showLabel is false', () => {
    const { getByTestId, getByRole } = rendersSegmentedControl();

    fireEvent.mouseOver(getByRole('button', { name: 'Gallery' }));
    expect(getByTestId('tooltip-content')).toHaveTextContent('Gallery');
  });

  it('renders disabled option correctly', () => {
    const { getByRole } = render(
      <ProvidedSegmentedControl
        dataTestId='segmented-control'
        selectedValue='list'
      >
        <SegmentedControl.Option iconCode='list' label='List' value='list' />
        <SegmentedControl.Option
          disabled
          iconCode='grid_view'
          label='Gallery'
          value='gallery'
        />
      </ProvidedSegmentedControl>
    );

    const galleryOption = getByRole('button', { name: 'Gallery' });
    expect(galleryOption).toBeDisabled();
  });

  it('renders label when icon is not present even with showLabel flag as false', () => {
    const { getByRole } = render(
      <ProvidedSegmentedControl
        dataTestId='segmented-control'
        selectedValue='list'
        showLabel={false}
      >
        <SegmentedControl.Option label='List' value='list' />
      </ProvidedSegmentedControl>
    );

    expect(getByRole('button', { name: 'List' })).toHaveTextContent('List');
  });
});

const rendersSegmentedControl = (handleChange?: () => void) => {
  return render(
    <ProvidedSegmentedControl
      dataTestId='segmented-control'
      {...(handleChange && { onChangeControl: handleChange })}
    >
      <SegmentedControl.Option iconCode='list' label='List' value='list' />
      <SegmentedControl.Option
        iconCode='grid_view'
        label='Gallery'
        value='gallery'
      />
    </ProvidedSegmentedControl>
  );
};
