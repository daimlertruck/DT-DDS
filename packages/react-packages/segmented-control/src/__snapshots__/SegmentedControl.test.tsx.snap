// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SegmentedControl renders without errors 1`] = `
.emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 4px;
  border: 1px solid #DEDEDE;
  border-radius: 0px;
}

.emotion-2 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
}

.emotion-4 {
  font-weight: 400;
  font-size: 0.875rem;
  line-height: 1.25rem;
  letter-spacing: normal;
  margin: 0;
  text-transform: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 4px;
  cursor: pointer;
  background-color: #FFFFFF;
  color: #292929;
  border: none;
  border-radius: 0px;
  padding: 4px 4px;
}

.emotion-4:not(:disabled):hover {
  background-color: #DEDEDE;
  color: #292929;
}

.emotion-4:not(:disabled):hover i {
  color: #131313;
}

.emotion-4:not(:disabled)[aria-pressed="true"] {
  background-color: #000000;
  color: #FFFFFF;
}

.emotion-4:not(:disabled)[aria-pressed="true"] i {
  color: #FFFFFF;
}

.emotion-4:disabled {
  color: #A3A3A3;
  cursor: not-allowed;
}

.emotion-4:disabled i {
  color: #A3A3A3;
}

.emotion-6 {
  font-family: DTUI-icons-outlined;
  font-weight: normal;
  font-style: normal;
  font-size: 16px;
  display: inline-block;
  line-height: 1;
  text-transform: none;
  letter-spacing: normal;
  word-wrap: normal;
  white-space: nowrap;
  font-variation-settings: 'FILL' 0;
  color: #292929;
  -webkit-font-smoothing: antialiased;
  text-rendering: optimizeLegibility;
  -moz-osx-font-smoothing: grayscale;
  font-feature-settings: 'liga';
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.emotion-6:hover,
.emotion-6:active {
  color: #00677F;
}

<div>
  <div
    class="emotion-0 emotion-1"
    data-testid="segmented-control"
    role="group"
  >
    <div
      class="emotion-2 emotion-3"
      data-testid="tooltip-container"
    >
      <button
        aria-label="List"
        aria-pressed="false"
        class="emotion-4 emotion-5"
        type="button"
      >
        <i
          class="emotion-6 emotion-7"
          color="#292929"
          data-testid="icon"
        >
          list
        </i>
      </button>
    </div>
    <div
      class="emotion-2 emotion-3"
      data-testid="tooltip-container"
    >
      <button
        aria-label="Gallery"
        aria-pressed="true"
        class="emotion-4 emotion-5"
        type="button"
      >
        <i
          class="emotion-6 emotion-7"
          color="#292929"
          data-testid="icon"
        >
          grid_view
        </i>
      </button>
    </div>
  </div>
</div>
`;
